pkg/apis/core/v1alpha1/projectroletemplate_types.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2020 The Bulward Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ProjectRoleTemplateSpec describes the desired state of ProjectRoleTemplate.
+type ProjectRoleTemplateSpec struct {
+	// Metadata	contains additional human readable ProjectRoleTemplate details.
+	Metadata *ProjectRoleTemplateMetadata `json:"metadata,omitempty"`
+	// BindTo defines the member types of the Project that this ProjectRoleTemplate will be bound to.
+	BindTo []BindingType `json:"bindTo,omitempty"`
+	// ProjectSelector selects Project objects that this ProjectRoleTemplate should be applied to.
+	ProjectSelector *metav1.LabelSelector `json:"projectSelector,omitempty"`
+	// Rules defines the Role that this ProjectRoleTemplate refers to.
+	Rules []rbacv1.PolicyRule `json:"rules"`
+}
+
+// ProjectRoleTemplateMetadata contains the metadata of the ProjectRoleTemplate.
+type ProjectRoleTemplateMetadata struct {
+	// DisplayName is the human-readable name of this ProjectRoleTemplate.
+	// +kubebuilder:validation:MinLength=1
+	DisplayName string `json:"displayName"`
+	// Description is the long and detailed description of the ProjectRoleTemplate.
+	// +kubebuilder:validation:MinLength=1
```

Comment:
> well...now that I'm playing around with kubecarrier a bit more...do we really want enforcing all our objects have Description and DisplayName?
We are not really enforcing that, because metadata is `omitempty` in the Spec.

> What I'm suggesting is allowing Description/ShortDescription being empty and default to sensible defaults (i.e. DisplayName being named if empty, shortDescription being first X chars from Description if empty, etc.)

Yeah, we can do that, but we need to implement some webhook logic for that I think, from start, I would like to keep everything simple.

Reply:
ok, simple and easy to start.